src/base1.rs

```diff
@@ -0,0 +1,1985 @@
+    /// Reference to a node with the given key, if found.
+    ///
+    /// If this is `Some` then it will point to the same node as `right[0]`.
+    found: Option<&'a Node<K, V>>,
+
+    /// Adjacent nodes with smaller keys (predecessors).
+    left: [&'a Tower<K, V>; MAX_HEIGHT],
+
+    /// Adjacent nodes with equal or greater keys (successors).
+    right: [Shared<'a, Node<K, V>>; MAX_HEIGHT],
+}
+
+/// Frequently modified data associated with a skip list.
+struct HotData {
+    /// The seed for random height generation.
+    seed: AtomicUsize,
+
+    /// The number of entries in the skip list.
+    len: AtomicUsize,
+
+    /// Highest tower currently in use. This value is used as a hint for where
+    /// to start lookups and never decreases.
+    max_height: AtomicUsize,
+}
+
+/// A lock-free skip list.
+// TODO(stjepang): Embed a custom `epoch::Collector` inside `SkipList<K, V>`. Instead of adding
+// garbage to the default global collector, we should add it to a local collector tied to the
+// particular skip list instance.
+//
+// Since global collector might destroy garbage arbitrarily late in the future, some skip list
+// methods have `K: 'static` and `V: 'static` bounds. But a local collector embedded in the skip
+// list would destroy all remaining garbage when the skip list is dropped, so in that case we'd be
+// able to remove those bounds on types `K` and `V`.
+//
+// As a further future optimization, if `!mem::needs_drop::<K>() && !mem::needs_drop::<V>()`
+// (neither key nor the value have destructors), there's no point in creating a new local
+// collector, so we should simply use the global one.
+pub struct SkipList<K, V> {
+    /// The head of the skip list (just a dummy node, not a real entry).
+    head: Head<K, V>,
+
+    /// The `Collector` associated with this skip list.
+    collector: Collector,
+
+    /// Hot data associated with the skip list, stored in a dedicated cache line.
+    hot_data: CachePadded<HotData>,
+}
+
+unsafe impl<K: Send + Sync, V: Send + Sync> Send for SkipList<K, V> {}
+unsafe impl<K: Send + Sync, V: Send + Sync> Sync for SkipList<K, V> {}
+
+impl<K, V> SkipList<K, V> {
+    /// Returns a new, empty skip list.
+    pub fn new(collector: Collector) -> SkipList<K, V> {
+        SkipList {
+            head: Head::new(),
+            collector,
+            hot_data: CachePadded::new(HotData {
+                seed: AtomicUsize::new(1),
+                len: AtomicUsize::new(0),
+                max_height: AtomicUsize::new(1),
+            }),
+        }
+    }
+
+    /// Returns `true` if the skip list is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the number of entries in the skip list.
+    ///
+    /// If the skip list is being concurrently modified, consider the returned number just an
+    /// approximation without any guarantees.
+    pub fn len(&self) -> usize {
+        self.hot_data.len.load(Ordering::Relaxed)
```

Comment:
Since `len` is just an approximation updated with `Relaxed` ordering and could sometimes be a little off, I fear concurrent inserts and deletes might sometimes make the number fall below zero, which would make it a huge number due to underflow.

The fix is easy:

```rust
let len = self.hot_data.len.load(Ordering::Relaxed);
if len > isize::MAX as usize {
    0
} else {
    len
}
```